server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -102,14 +122,137 @@ public void writeFileChunk(
         int totalTranslogOps,
         ActionListener<Void> listener
     ) {
-        // TODO
+        try {
+            multiFileWriter.writeFileChunk(metadata, position, content, lastChunk);
+            listener.onResponse(null);
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
     }
 
     /**
      * Start the Replication event.
      * @param listener {@link ActionListener} listener.
      */
     public void startReplication(ActionListener<Void> listener) {
-        // TODO
+        state.setStage(SegmentReplicationState.Stage.REPLICATING);
+        final StepListener<CheckpointInfoResponse> checkpointInfoListener = new StepListener<>();
+        final StepListener<GetSegmentFilesResponse> getFilesListener = new StepListener<>();
+        final StepListener<Void> finalizeListener = new StepListener<>();
+
+        // Get list of files to copy from this checkpoint.
+        source.getCheckpointMetadata(getId(), checkpoint, checkpointInfoListener);
+
+        checkpointInfoListener.whenComplete(checkpointInfo -> getFiles(checkpointInfo, getFilesListener), listener::onFailure);
+        getFilesListener.whenComplete(
+            response -> finalizeReplication(checkpointInfoListener.result(), finalizeListener),
+            listener::onFailure
+        );
+        finalizeListener.whenComplete(r -> listener.onResponse(null), listener::onFailure);
+    }
+
+    private void getFiles(CheckpointInfoResponse checkpointInfo, StepListener<GetSegmentFilesResponse> getFilesListener)
+        throws IOException {
+        final Store.MetadataSnapshot snapshot = checkpointInfo.getSnapshot();
+        Store.MetadataSnapshot localMetadata = getMetadataSnapshot();
+        final Store.RecoveryDiff diff = snapshot.recoveryDiff(localMetadata);
+        logger.debug("Replication diff {}", diff);
+        if (diff.different.isEmpty() == false) {
```

Comment:
I think we should add a comment with a bit of context here on why we are throwing an error and failing the shard when the replica has any segments of the same name that differ from the incoming snapshot.